[Issue 277] add cloud_init_disable_upgrade_packages option

mpywell opened the pull request:
Add option to disable the `Upgrade Packages` setting on Cloud-Init Drives attached to VMs.

`Upgrade Packages` is introduced in Proxmox VE 8 and logic is added to only configure this option on backend versions that are compatible.

Closes #277

VadimLakt committed:
fix for remove the old cloud-init on local disks

mpywell committed:
common: add skip_convert_to_template option

mpywell committed:
skip_convert_to_template: return a running VM, update force flag to work with VM builds

mpywell committed:
common: skip_convert_to_template updates
- updated VM stop/start logic for hardware changes to a VM build
- reverted getExistingTemplate signature changes

mpywell committed:
fix error message

mpywell committed:
common: add cloud_init_disable_upgrade_packages

mpywell committed:
common: add cloud_init_disable_upgrade_packages

mpywell committed:
cloud_init_disable_upgrade_packages: update ui output when incompatible

mpywell committed:
common: keep generated ISOs when unmount is false

mogrogan committed:
version: prepare v1.2.3-dev

mogrogan committed:
version: cut release v1.2.2

itsmeow committed:
Broaden check for duplicate VM IDs

dependabot[bot] committed:
build(deps): bump github.com/hashicorp/packer-plugin-sdk

Bumps [github.com/hashicorp/packer-plugin-sdk](https://github.com/hashicorp/packer-plugin-sdk) from 0.5.4 to 0.6.1.
- [Release notes](https://github.com/hashicorp/packer-plugin-sdk/releases)
- [Changelog](https://github.com/hashicorp/packer-plugin-sdk/blob/main/CHANGELOG.md)
- [Commits](https://github.com/hashicorp/packer-plugin-sdk/compare/v0.5.4...v0.6.1)

---
updated-dependencies:
- dependency-name: github.com/hashicorp/packer-plugin-sdk
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

lbajolet-hashicorp committed:
github: bump golangci-lint to v1.60.1

Version 1.54.2 of the golangci-lint tool is not compatible with go
1.23, which is the version the project is now compiled with.

In order for CI to be able to run correctly, we are therefore bumping
the golangci-lint dependency to 1.60.1 now, so that the lint job can
run again.

lbajolet-hashicorp committed:
golangci-lint: update config file to conform with 1.61.0 changes

A series of changes have made the current .golangci.yml configuration
file non-conformant with latest updates, causing warnings and errors
when attempting to run the tool on the plugin.

This commit fixes all the reported issues with the configurtion,
removing dead checkers and updating the deprecated/removed options in
the config file.

devashish-patel committed:
Merge pull request #322 from hashicorp/update-pr-template-pci

Update PR template for PCI

maksimsamt committed:
Merge pull request #308 from maksimsamt/main

[Issue 307] fix missing CpuType

hashicorp-copywrite[bot] committed:
[COMPLIANCE] Update Copyright and License Headers (Batch 1 of 1)

njuettner committed:
Respect proxy environment variables for Proxmox API connections

When no explicit proxy is configured, the Proxmox API client creates an
http.Transport with Proxy: nil (via the Telmate library), which
explicitly disables Go's default proxy behavior. This means standard
proxy environment variables (HTTPS_PROXY, HTTP_PROXY, NO_PROXY) are
completely ignored.

Fix this by creating an http.Client with http.ProxyFromEnvironment and
passing it to the Telmate library's NewClient(), so proxy env vars are
respected. When no proxy env vars are set, behavior is unchanged.

Fixes #346

lloydmcl committed:
[Issue 342] Use VM name as fallback when looking up existing template with -force

When template_name is not explicitly set and -force is used,
getExistingTemplate() searched for VMs with an empty name string,
finding nothing and silently skipping deletion of the old template.

This aligns the lookup behaviour with stepFinalizeTemplateConfig,
which already falls back to VMName when TemplateName is empty.

lloydmcl committed:
[Issues 312, 332] Mark required fields in docs, add TPM config documentation, fix TPM comment

The proxmox_url, username, and node fields were documented as optional
but validated as required in Prepare(). Added required:"true" struct
tags so packer-sdc generates them into Config-required.mdx.

TPM config doc partials existed but were not included in either builder
page. Added @include directives to both iso.mdx and clone.mdx.

Fixed copy-paste error in tpmConfig.TPMStoragePool comment which
incorrectly referenced "EFI disk" instead of "TPM state".

CaptaiNiveau commented:
Just started implementing this myself, until I stumbled upon this PR. @JenGoldstrich [you mentioned over here](https://github.com/hashicorp/packer-plugin-proxmox/pull/308#pullrequestreview-3064628944) that you wanted to discuss maintenance status of this plugin with the team - what's the current state of that? It'd be nice if small things like this PR were approved.

I need this option, because otherwise Arch Linux automatically upgrades its packages in the iso and immediately fills up its cowspace. Yes, I can increase the size of the cowspace afterwards, but just disabling auto upgrade is obviously the significantly better fix.

Thanks for all the nice IaC tooling, would love to see it improve :)

CaptaiNiveau commented:
Oh btw, I rebased that PR and fixed the small merge conflicts [over here](https://github.com/CaptaiNiveau/packer-plugin-proxmox/tree/feat/277_cloud_init_upgrade_packages). I'll be testing it out in the next few days but I doubt that there will be any issues.

CaptaiNiveau commented:
Hm, sadly it doesn't seem to work as expected yet. Just installed the plugin, but swapping it on/off doesn't change anything about the `upgrade packages` flag in Proxmox Cloud Init. Hm, I'll try to debug this.

CaptaiNiveau commented:
Ah, that PR actually only focused on the resulting template. I expanded on it and also use this variable if the vm is cloned instead of the iso. That behavior is in line with other existing variables, like the ssh public key. That one is also pulled in from the common communicator module into the Cloud Init configuration of the cloned vm.

mpywell commented:
Hey @CaptaiNiveau good catch, I must have missed adding this for the clone builder back when I submitted this PR 2 years ago. I do hope this repo gets some attention, I lost momentum as each PR I wrote had a long wait before being reviewed, some like this one are still awaiting an initial review.

hashicorp-copywrite[bot] committed:
[COMPLIANCE] Add/Update Copyright Headers

tanmay-hc commented on builder/proxmox/common/step_finalize_config.go:
```diff
@@ -22,6 +22,7 @@ type stepFinalizeTemplateConfig struct{}
 type templateFinalizer interface {
 	GetVmConfig(*proxmox.VmRef) (map[string]interface{}, error)
 	SetVmConfig(*proxmox.VmRef, map[string]interface{}) (interface{}, error)
+	Version() (proxmox.Version, error)
```
This introduces a new `version` dependency on every finalize path, even when `cloud_init` is off. That changes existing behavior and can now fail builds unrelated to this option. Can we move the version lookup into the exact branch where we actually need to decide whether to write `ciupgrade`?

tanmay-hc commented on builder/proxmox/common/step_finalize_config.go:
```diff
@@ -87,11 +96,25 @@ func (s *stepFinalizeTemplateConfig) Run(ctx context.Context, state multistep.St
+						if c.CloudInitDisableUpgradePackages {
+							ui.Say("cloud_init_disable_upgrade_packages is set to true, but not supported in Proxmox versions lower than 8. No changes made.")
+						}
+					}
```
This also changes the default behavior: when the new field is omitted, we still write `ciupgrade=true`. That means existing templates/clones with `ciupgrade=false` can get flipped on upgrade. We probably need tri-state semantics here, or to only send `ciupgrade` when the user explicitly set the option.

tanmay-hc reviewed (changes requested):
I know the review is a bit too late. Apologies for that.
Please address the comments and do a rebase.
We should be good to merge.

tanmay-hc committed:
test: add missing GetVmState mock for template_name fallback force-delete case

Reconciles PR #283's new GetVmState() call in the force-delete path with
the TestStartVMWithForce case added on main (#342), which lacked a
mockGetVmState and panicked after the merge.

mpywell committed:
reconfigure cloud_init_disable_upgrade_packages as trilean, add functionality for clone builder, move proxmox version check to occur only if trilean is set

mpywell committed:
regenerate documentation and specs

mpywell committed:
Merge branch 'main' into feat/277_cloud_init_upgrade_packages

mpywell commented:
Hi @tanmay-hc, have rebased and made some adjustments.

- CloudInitDisableUpgradePackages is now Trilean
- Version is only called if the Trilean has a set value
- I've added the functionality to the clone builder to resolve @CaptaiNiveau's earlier comments
